In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df= pd.read_csv('/content/drive/MyDrive/Projects/House_price/train.csv')
df

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

In [3]:
df.shape

(1460, 81)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [5]:
numeric_col = df.select_dtypes(include=['int64','float64']).columns
numeric_col

Index(['Id', 'MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual',
       'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF',
       'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea',
       'MiscVal', 'MoSold', 'YrSold', 'SalePrice'],
      dtype='object')

In [6]:
df[numeric_col]= df[numeric_col].fillna(df[numeric_col].mean())

In [7]:
categorical_col= df.select_dtypes(include=['object']).columns
categorical_col

Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')

In [8]:
corr= df[numeric_col].corr()
corr

Id  MSSubClass  LotFrontage   LotArea  OverallQual  \
Id             1.000000    0.011156    -0.009601 -0.033226    -0.028365   
MSSubClass     0.011156    1.000000    -0.357056 -0.139781     0.032628   
LotFrontage   -0.009601   -0.357056     1.000000  0.306795     0.234196   
LotArea       -0.033226   -0.139781     0.306795  1.000000     0.105806   
OverallQual   -0.028365    0.032628     0.234196  0.105806     1.000000   
OverallCond    0.012609   -0.059316    -0.052820 -0.005636    -0.091932   
YearBuilt     -0.012713    0.027850     0.117598  0.014228     0.572323   
YearRemodAdd  -0.021998    0.040581     0.082746  0.013788     0.550684   
MasVnrArea    -0.050199    0.022895     0.179283  0.103960     0.410238   
BsmtFinSF1    -0.005024   -0.069836     0.215828  0.214103     0.239666   
BsmtFinSF2    -0.005968   -0.065649     0.043340  0.111170    -0.059119   
BsmtUnfSF     -0.007940   -0.140759     0.122156 -0.002618     0.308159   
TotalBsmtSF   -0.015415   -0.238518     0.363358  0.260833     0.537808   
1stFlrSF       0.010496   -0.251758     0.414266  0.299475     0.476224   
2ndFlrSF       0.005590    0.307886     0.072483  0.050986     0.295493   
LowQualFinSF  -0.044230    0.046474     0.036849  0.004779    -0.030429   
GrLivArea      0.008273    0.074853     0.368392  0.263116     0.593007   
BsmtFullBath   0.002289    0.003491     0.091481  0.158155     0.111098   
BsmtHalfBath  -0.020155   -0.002333    -0.006419  0.048046    -0.040150   
FullBath       0.005587    0.131608     0.180424  0.126031     0.550600   
HalfBath       0.006784    0.177354     0.048258  0.014259     0.273458   
BedroomAbvGr   0.037719   -0.023438     0.237023  0.119690     0.101676   
KitchenAbvGr   0.002951    0.281721    -0.005805 -0.017784    -0.183882   
TotRmsAbvGrd   0.027239    0.040380     0.320146  0.190015     0.427452   
Fireplaces    -0.019772   -0.045569     0.235755  0.271364     0.396765   
GarageYrBlt    0.000070    0.080187     0.064324 -0.024812     0.518018   
GarageCars     0.016570   -0.040110     0.269729  0.154871     0.600671   
GarageArea     0.017634   -0.098672     0.323663  0.180403     0.562022   
WoodDeckSF    -0.029643   -0.012579     0.077106  0.171698     0.238923   
OpenPorchSF   -0.000477   -0.006100     0.137454  0.084774     0.308819   
EnclosedPorch  0.002889   -0.012037     0.009790 -0.018340    -0.113937   
3SsnPorch     -0.046635   -0.043825     0.062335  0.020423     0.030371   
ScreenPorch    0.001330   -0.026030     0.037684  0.043160     0.064886   
PoolArea       0.057044    0.008283     0.180868  0.077672     0.065166   
MiscVal       -0.006242   -0.007683     0.001168  0.038068    -0.031406   
MoSold         0.021172   -0.013585     0.010158  0.001205     0.070815   
YrSold         0.000712   -0.021407     0.006768 -0.014261    -0.027347   
SalePrice     -0.021917   -0.084284     0.334901  0.263843     0.790982   

               OverallCond  YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  \
Id                0.012609  -0.012713     -0.021998   -0.050199   -0.005024   
MSSubClass       -0.059316   0.027850      0.040581    0.022895   -0.069836   
LotFrontage      -0.052820   0.117598      0.082746    0.179283    0.215828   
LotArea          -0.005636   0.014228      0.013788    0.103960    0.214103   
OverallQual      -0.091932   0.572323      0.550684    0.410238    0.239666   
OverallCond       1.000000  -0.375983      0.073741   -0.127788   -0.046231   
YearBuilt        -0.375983   1.000000      0.592855    0.314745    0.249503   
YearRemodAdd      0.073741   0.592855      1.000000    0.179186    0.128451   
MasVnrArea       -0.127788   0.314745      0.179186    1.000000    0.263582   
BsmtFinSF1       -0.046231   0.249503      0.128451    0.263582    1.000000   
BsmtFinSF2        0.040229  -0.049107     -0.067759   -0.072302   -0.050117   
BsmtUnfSF        -0.136841   0.149040      0.181133    0.114184   -0.495251   
TotalBsmtSF      -0.171098   0.391452      0.291066    0.362452    0

In [9]:
df[categorical_col]= df[categorical_col].fillna('NA')

In [10]:
df[categorical_col].nunique()

MSZoning          5
Street            2
Alley             3
LotShape          4
LandContour       4
Utilities         2
LotConfig         5
LandSlope         3
Neighborhood     25
Condition1        9
Condition2        8
BldgType          5
HouseStyle        8
RoofStyle         6
RoofMatl          8
Exterior1st      15
Exterior2nd      16
MasVnrType        4
ExterQual         4
ExterCond         5
Foundation        6
BsmtQual          5
BsmtCond          5
BsmtExposure      5
BsmtFinType1      7
BsmtFinType2      7
Heating           6
HeatingQC         5
CentralAir        2
Electrical        6
KitchenQual       4
Functional        7
FireplaceQu       6
GarageType        7
GarageFinish      4
GarageQual        6
GarageCond        6
PavedDrive        3
PoolQC            4
Fence             5
MiscFeature       5
SaleType          9
SaleCondition     6
dtype: int64

In [11]:
from sklearn.preprocessing import OneHotEncoder

In [12]:
encoder= OneHotEncoder(handle_unknown='ignore')

In [13]:
encoder.fit(df[categorical_col])

OneHotEncoder(handle_unknown='ignore')

In [14]:
encoded_col= list(encoder.get_feature_names_out(categorical_col))
print(encoded_col)

['MSZoning_C (all)', 'MSZoning_FV', 'MSZoning_RH', 'MSZoning_RL', 'MSZoning_RM', 'Street_Grvl', 'Street_Pave', 'Alley_Grvl', 'Alley_NA', 'Alley_Pave', 'LotShape_IR1', 'LotShape_IR2', 'LotShape_IR3', 'LotShape_Reg', 'LandContour_Bnk', 'LandContour_HLS', 'LandContour_Low', 'LandContour_Lvl', 'Utilities_AllPub', 'Utilities_NoSeWa', 'LotConfig_Corner', 'LotConfig_CulDSac', 'LotConfig_FR2', 'LotConfig_FR3', 'LotConfig_Inside', 'LandSlope_Gtl', 'LandSlope_Mod', 'LandSlope_Sev', 'Neighborhood_Blmngtn', 'Neighborhood_Blueste', 'Neighborhood_BrDale', 'Neighborhood_BrkSide', 'Neighborhood_ClearCr', 'Neighborhood_CollgCr', 'Neighborhood_Crawfor', 'Neighborhood_Edwards', 'Neighborhood_Gilbert', 'Neighborhood_IDOTRR', 'Neighborhood_MeadowV', 'Neighborhood_Mitchel', 'Neighborhood_NAmes', 'Neighborhood_NPkVill', 'Neighborhood_NWAmes', 'Neighborhood_NoRidge', 'Neighborhood_NridgHt', 'Neighborhood_OldTown', 'Neighborhood_SWISU', 'Neighborhood_Sawyer', 'Neighborhood_SawyerW', 'Neighborhood_Somerst', 'Ne

In [15]:
df[encoded_col]= encoder.transform(df[categorical_col]).toarray()

/tmp/ipython-input-3593936961.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[encoded_col]= encoder.transform(df[categorical_col]).toarray()
/tmp/ipython-input-3593936961.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  df[encoded_col]= encoder.transform(df[categorical_col]).toarray()
/tmp/ipython-input-3593936961.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(ax

In [16]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave    NA      Reg   
1   2          20       RL         80.0     9600   Pave    NA      Reg   
2   3          60       RL         68.0    11250   Pave    NA      IR1   
3   4          70       RL         60.0     9550   Pave    NA      IR1   
4   5          60       RL         84.0    14260   Pave    NA      IR1   

  LandContour Utilities  ... SaleType_ConLw SaleType_New SaleType_Oth  \
0         Lvl    AllPub  ...            0.0          0.0          0.0   
1         Lvl    AllPub  ...            0.0          0.0          0.0   
2         Lvl    AllPub  ...            0.0          0.0          0.0   
3         Lvl    AllPub  ...            0.0          0.0          0.0   
4         Lvl    AllPub  ...            0.0          0.0          0.0   

  SaleType_WD SaleCondition_Abnorml SaleCondition_AdjLand  \
0         1.0                   0.0                   0.0   
1         1.0                   0.0                   0.0   
2         1.0                   0.0                   0.0   
3         1.0                   1.0                   0.0   
4         1.0                   0.0                   0.0   

  SaleCondition_Alloca  SaleCondition_Family  SaleCondition_Normal  \
0                  0.0                   0.0                   1.0   
1                  0.0                   0.0                   1.0   
2                  0.0                   0.0                   1.0   
3                  0.0                   0.0                   0.0   
4                  0.0                   0.0                   1.0   

   SaleCondition_Partial  
0                    0.0  
1                    0.0  
2                    0.0  
3                    0.0  
4                    0.0  

[5 rows x 348 columns]

In [17]:
from sklearn.model_selection import train_test_split

In [18]:
val_df, train_df= train_test_split(df, test_size=0.8, random_state=42)

In [19]:
train_df

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
892    893          20       RL    70.000000     8414   Pave    NA      Reg   
1105  1106          60       RL    98.000000    12256   Pave    NA      IR1   
413    414          30       RM    56.000000     8960   Pave  Grvl      Reg   
522    523          50       RM    50.000000     5000   Pave    NA      Reg   
1036  1037          20       RL    89.000000    12898   Pave    NA      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1069  1070          45       RL    60.000000     9600   Pave    NA      Reg   
476    477          20       RL    75.000000    13125   Pave    NA      Reg   
157    158          60       RL    92.000000    12003   Pave    NA      Reg   
16      17          20       RL    70.049958    11241   Pave    NA      IR1   
1072  1073          50       RL    50.000000     7585   Pave    NA      Reg   

     LandContour Utilities  ... SaleType_ConLw SaleType_New SaleType_Oth  \
892          Lvl    AllPub  ...            0.0          0.0          0.0   
1105         Lvl    AllPub  ...            0.0          0.0          0.0   
413          Lvl    AllPub  ...            0.0          0.0          0.0   
522          Lvl    AllPub  ...            0.0          0.0          0.0   
1036         HLS    AllPub  ...            0.0          0.0          0.0   
...          ...       ...  ...            ...          ...          ...   
1069         Lvl    AllPub  ...            0.0          0.0          0.0   
476          Lvl    AllPub  ...            0.0          0.0          0.0   
157          Lvl    AllPub  ...            0.0          1.0          0.0   
16           Lvl    AllPub  ...            0.0          0.0          0.0   
1072         Lvl    AllPub  ...            0.0          0.0          0.0   

     SaleType_WD SaleCondition_Abnorml SaleCondition_AdjLand  \
892          1.0                   0.0                   0.0   
1105         1.0                   0.0                   0.0   
413          1.0                   0.0                   0.0   
522          1.0                   0.0                   0.0   
1036         1.0                   0.0                   0.0   
...          ...                   ...                   ...   
1069         1.0                   0.0                   0.0   
476          1.0                   0.0                   0.0   
157          0.0                   0.0                   0.0   
16           1.0                   0.0                   0.0   
1072         1.0                   0.0                   0.0   

     SaleCondition_Alloca  SaleCondition_Family  SaleCondition_Normal  \
892                   0.0                   0.0                   1.0   
1105                  0.0                   0.0                   1.0   
413                   0.0                   0.0                   1.0   
522                   0.0                   0.0                   1.0   
1036                  0.0                   0.0                   1.0   
...                   ...                   ...                   ...   
1069                  0.0                   0.0                   1.0   
476                   0.0                   0.0                   1.0   
157                   0.0                   0.0                   0.0   
16                    0.0                   0.0                   1.0   
1072                  0.0                   0.0                   1.0   

      SaleCondition_Partial  
892                     0.0  
1105                    0.0  
413                     0.0  
522                     0.0  
1036                    0.0  
...                     ...  
1069                    0.0  
476                     0.0  
157                     1.0  
16                      0.0  
1072                    0.0  

[1168 rows x 348 columns]

In [20]:
input_col= list(numeric_col.drop('SalePrice')) + encoded_col
input_col

['Id',
 'MSSubClass',
 'LotFrontage',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'YearRemodAdd',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'BsmtHalfBath',
 'FullBath',
 'HalfBath',
 'BedroomAbvGr',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageYrBlt',
 'GarageCars',
 'GarageArea',
 'WoodDeckSF',
 'OpenPorchSF',
 'EnclosedPorch',
 '3SsnPorch',
 'ScreenPorch',
 'PoolArea',
 'MiscVal',
 'MoSold',
 'YrSold',
 'MSZoning_C (all)',
 'MSZoning_FV',
 'MSZoning_RH',
 'MSZoning_RL',
 'MSZoning_RM',
 'Street_Grvl',
 'Street_Pave',
 'Alley_Grvl',
 'Alley_NA',
 'Alley_Pave',
 'LotShape_IR1',
 'LotShape_IR2',
 'LotShape_IR3',
 'LotShape_Reg',
 'LandContour_Bnk',
 'LandContour_HLS',
 'LandContour_Low',
 'LandContour_Lvl',
 'Utilities_AllPub',
 'Utilities_NoSeWa',
 'LotConfig_Corner',
 'LotConfig_CulDSac',
 'LotConfig_FR2',
 'LotConfig_FR3',
 'LotConfig_Inside',
 'LandS

In [21]:
from sklearn.preprocessing import MinMaxScaler

In [22]:
scaler= MinMaxScaler().fit(train_df[numeric_col.drop('SalePrice')])

In [23]:
train_df[numeric_col.drop('SalePrice')]= scaler.transform(train_df[numeric_col.drop('SalePrice')])
val_df[numeric_col.drop('SalePrice')]= scaler.transform(val_df[numeric_col.drop('SalePrice')])

In [24]:
def return_mean(inputs):
    return np.full(len(inputs), train_df.SalePrice.mean())

In [25]:
mean_preds = return_mean(train_df[input_col])


In [26]:
from sklearn.metrics import mean_absolute_error
target_col = 'SalePrice'

In [27]:
mean_absolute_error(mean_preds, train_df[target_col])

58266.75105554512

Linear Model


In [28]:
from sklearn.linear_model import LinearRegression

In [29]:
lin_model= LinearRegression()


In [30]:
linear_model= lin_model.fit(train_df[input_col], train_df[target_col])

In [31]:
linreg_train_pred= lin_model.predict(train_df[input_col])
linreg_val_pred= lin_model.predict(val_df[input_col])

In [32]:
mean_absolute_error(linreg_train_pred, train_df[target_col])

13019.306276780617

In [33]:
mean_absolute_error(linreg_val_pred, val_df[target_col])

17174.892591085463

Decision Tree

In [34]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

In [35]:
dectree_model= RandomForestRegressor(n_estimators=100, random_state=42)

In [36]:
dectree_model.fit(train_df[input_col], train_df[target_col])

RandomForestRegressor(random_state=42)

In [37]:
dectree_train_pred=dectree_model.predict(train_df[input_col])
dectree_val_pred= dectree_model.predict(val_df[input_col])

In [38]:
mean_absolute_error(dectree_train_pred, train_df[target_col])

6872.617491438356

In [39]:
mean_absolute_error(dectree_val_pred, val_df[target_col])


15254.355102739724

Gradirnt Boosted

In [40]:
!pip install xgboost

In [41]:
from xgboost import XGBRegressor

In [42]:
xg_model= XGBRegressor(n_estimators=10, n_jobs=-1, random_state=42)

In [43]:
xg_model.fit(train_df[input_col], train_df[target_col])

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             feature_weights=None, gamma=None, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=None, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=None,
             max_leaves=None, min_child_weight=None, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=10,
             n_jobs=-1, num_parallel_tree=None, ...)

In [44]:
xg_train_pred=dectree_model.predict(train_df[input_col])
xg_val_pred= dectree_model.predict(val_df[input_col])

In [45]:
mean_absolute_error(xg_train_pred, train_df[target_col])

6872.617491438356

In [46]:
mean_absolute_error(xg_val_pred, val_df[target_col])

15254.355102739724

So our final model is Random Forest Model which reduced error from mean prediction of 58266 to  validation error of 15254.